src/main/java/org/aksw/simba/lemming/creation/Inferer.java

```diff
@@ -198,70 +216,89 @@ public void iterateStmts(Model newModel, Model sourceModel, OntModel ontModel, M
 			List<? extends OntResource> domain = property.listDomain().toList();
 			for (OntResource curResource : domain) {
 				Statement subjType = ResourceFactory.createStatement(subject, RDF.type, curResource);
-				if (!newModel.contains(subjType)) {
-					newStmts.add(subjType);
-				}
+				newStmts.add(subjType);
+
 			}
 			if (object.isResource()) {
 				List<? extends OntResource> range = property.listRange().toList();
 				for (OntResource curResource : range) {
 					Statement objType = ResourceFactory.createStatement(object.asResource(), RDF.type, curResource);
-					if (!newModel.contains(objType)) {
-						newStmts.add(objType);
-					}
+					newStmts.add(objType);
 				}
 			}
 		}
 		return newStmts;
 	}
 
 	/**
-	 * Same as searchType(Statement statement, OntModel ontModel, Model newModel),
+	 * Same as searchType(Statement statement, Model model, OntModel ontModel),
 	 * but in our custom objects: For a given statement, this method searches for
 	 * the predicate of a model inside the Ontology. If found in the Ontology, it
 	 * then extracts the domain and range. Creating and adding a new triple with the
 	 * inferred type to the model.
 	 * 
 	 * @param statement  statement in which we want to check the predicate in the
 	 *                   ontology
-	 * @param newModel   where we add the new triples and therefore, where we check
+	 * @param model   where we add the new triples and therefore, where we check
 	 *                   if the statement is already existing in the model or not
-	 * @param uriNodeMap
+	 * @param iriEquiMap map each resource uri to its equivalent resources
 	 * @return a set of statements inferred from a property
 	 */
-	private Set<Statement> searchType(Statement statement, Model newModel, Map<String, Equivalent> uriNodeMap) {
+	private Set<Statement> searchType(Statement statement, Model model, Map<OntProperty, Set<OntProperty>> iriEquiMap) {
 		Set<Statement> newStmts = new HashSet<>();
 		Resource subject = statement.getSubject();
 		Property predicate = statement.getPredicate();
 		RDFNode object = statement.getObject();
 
-		Equivalent<OntProperty> node = uriNodeMap.get(predicate.toString());
+		Set<OntProperty> eqsSet = iriEquiMap.get(predicate);
 		OntProperty property = null;
+		if(eqsSet != null){
+			List<OntProperty> eqsProperties = new ArrayList<>(iriEquiMap.get(predicate));
+			property = eqsProperties.get(0);
+			//Add all domains and ranges of equivalent properties to property
+			for(int i=1; i<eqsProperties.size(); i++){
+				OntProperty currentProperty = eqsProperties.get(i);

```

Comment:
According to the `searchEquivalents` method, it is possible for the `iriEquiMap.get(predicate)` to be empty and your code will throw an exception in this case. 
From what I understand, I don't think you need to create a new list, you could just iterate the set directly and it would be safe to use if it's empty: 

```
Set<OntProperty> eqsProperties = iriEquiMap.get(predicate);
for(OntProperty currentProperty : eqsProperties){
	if(property == null) {
		property = currentProperty;
		continue;
	}
        ...
}
```

Reply:
Done

Reply:
I think this problem hasn't been solved, only the location changed.